stage0/gc.go

```diff
@@ -65,49 +67,111 @@ type mount struct {
 	id         int
 	parentId   int
 	mountPoint string
+	mode       string
 }
 
 type mounts []*mount
 
-// Less ensures that mounts are sorted in an order we can unmount; child before
-// parent. This works by pushing all child mounts up to the front. Parent
-// mounts are halted once they encounter a child. The actual sequential order
-// of the mounts doesn't matter; only that children are in front of parents.
-func (m mounts) Less(i, j int) bool { return m[i].id != m[j].parentId }
-func (m mounts) Len() int           { return len(m) }
-func (m mounts) Swap(i, j int)      { m[i], m[j] = m[j], m[i] }
+// getAncestors searches the list for all ancestors of the mount at index i
+func (m mounts) getAncestors(i int) map[int]struct{} {
+	ancestorMap := make(map[int]struct{}, 0)
+	current := m[i]
+	for found := true; found; {
+		found = false
+		for _, mnt := range m {
+			if mnt.id == current.parentId {
+				ancestorMap[mnt.id] = struct{}{}
+				current = mnt
+				found = true
+				break
+			}
+		}
+	}
+	return ancestorMap
+}
+
+// Less ensures that mounts are sorted in an order we can unmount; descendant before ancestor.
+// The requirement of transitivity for Less has to be fulfilled otherwise the sort algorithm will fail.
+func (m mounts) Less(i, j int) (result bool) {
+	iAncestors := m.getAncestors(i)
+	jAncestors := m.getAncestors(j)
+	result = result || (len(iAncestors) >= len(jAncestors))
+	return
+}
+func (m mounts) Len() int      { return len(m) }
+func (m mounts) Swap(i, j int) { m[i], m[j] = m[j], m[i] }
+
+// SafeOrder checks for transitivity and (un)mount order sanity
+func (m mounts) SafeOrder() error {
+	for i, mntCur := range m {
+		for j, mntNext := range m[(i + 1):] {
+			j := i + j
+			if !m.Less(i, j) {
+				return fmt.Errorf("Transitivity check failed for %d(%d) and %d(%d)", i, mntCur.id, j, mntNext.id)
+			}
+			if strings.HasPrefix(mntNext.mountPoint, mntCur.mountPoint) {
+				return fmt.Errorf("Must not unmount \n%q before\n%q.", mntCur.mountPoint, mntNext.mountPoint)
+			}
+		}
+	}
+	return nil
+}
 
 // getMountsForPrefix parses mi (/proc/PID/mountinfo) and returns mounts for path prefix
 func getMountsForPrefix(path string, mi io.Reader) (mounts, error) {
 	var podMounts mounts
 	sc := bufio.NewScanner(mi)
+	var (
+		mountId    int
+		parentId   int
+		mountPoint string
+		mode       string
+	)
+
+	lineRegex := regexp.MustCompile(`[^ ]+`)
 	for sc.Scan() {
-		var (
-			mountId    int
-			parentId   int
-			discard    string
-			mountPoint string
-		)
-
-		_, err := fmt.Sscanf(sc.Text(), "%d %d %s %s %s",
-			&mountId, &parentId, &discard, &discard, &mountPoint)
-		if err != nil {
-			return nil, err
+		line := sc.Text()
+		lineResult := lineRegex.FindAllStringSubmatch(line, -1)
```

Comment:
That works too. I had a different approach using regex that I simplified to basically just split by one or more spaces. I can switch it to using Split() if that's preferred.